### ISSUE TITLE
Reason for __interceptors being public?

### ISSUE THREAD

**Issue by martijnburgers** [castleproject/Core#260]
Hi there,

We are using dynamic proxies with Automapper a lot and came a cross a issue where Automapper is mapping the `__interceptors` field between two proxies. The reason for this is that Automapper maps every public field by default.

Is there a reason the `IInterceptor[] __interceptors` field on a proxy is public?

I can easily fix this with Automapper itself but just wanted to know if there is a valid reason for this field to be public.

**Comment by jonorossi** [comment]
Hi @martijnburgers, I didn't know the answer when you logged this issue and forgot all about it.

The history for why all static and instance fields on proxies are public would probably be a decision from 10+ years ago, but my guess is that it was required to run under partial trust which was common with shared hosting providers of the time.

I was interested to see what unit tests would fail if I changed it, and only the binary serialization ones failed until I made the 2nd change below:

```diff
diff --git a/src/Castle.Core/DynamicProxy/Generators/BaseProxyGenerator.cs b/src/Castle.Core/DynamicProxy/Generators/BaseProxyGenerator.cs
index 6ef8c5e..53289dd 100644
--- a/src/Castle.Core/DynamicProxy/Generators/BaseProxyGenerator.cs
+++ b/src/Castle.Core/DynamicProxy/Generators/BaseProxyGenerator.cs
@@ -159,7 +159,7 @@ namespace Castle.DynamicProxy.Generators
 
 		protected void CreateInterceptorsField(ClassEmitter emitter)
 		{
-			var interceptorsField = emitter.CreateField("__interceptors", typeof(IInterceptor[]));
+			var interceptorsField = emitter.CreateField("__interceptors", typeof(IInterceptor[]), FieldAttributes.Private);
 
 #if FEATURE_SERIALIZATION
 			emitter.DefineCustomAttributeFor<XmlIgnoreAttribute>(interceptorsField);
diff --git a/src/Castle.Core/DynamicProxy/Serialization/ProxyObjectReference.cs b/src/Castle.Core/DynamicProxy/Serialization/ProxyObjectReference.cs
index 49ad305..f0d6e4d 100644
--- a/src/Castle.Core/DynamicProxy/Serialization/ProxyObjectReference.cs
+++ b/src/Castle.Core/DynamicProxy/Serialization/ProxyObjectReference.cs
@@ -310,7 +310,7 @@ namespace Castle.DynamicProxy.Serialization
 
 		private void SetInterceptors(IInterceptor[] interceptors)
 		{
-			var interceptorField = proxy.GetType().GetField("__interceptors");
+			var interceptorField = proxy.GetType().GetField("__interceptors", BindingFlags.NonPublic | BindingFlags.Instance);
 			if (interceptorField == null)
 			{
 				throw new SerializationException(
```

I don't currently have any objections to making this type of change since partial trust is dead with .NET Core and barely used, however we would need to do it for all fields not just `__interceptors`.

**Comment by spencercw** [comment]
I've got a similar issue caused by the public __interceptors. Would it make sense to change the default from public to private in AbstractTypeEmitter.CreateField(string, Type, bool)? There don't seem to be many calls to CreateField() and I'm not sure any of them need to be public.

**Comment by fir3pho3nixx** [comment]
@spencercw - From what I am reading in this, it should be plausible. If you have more time, do you think it would be possible to package up a pull request? We would be happy to review a change here. We are just busy looking at other things.

**Comment by spencercw** [comment]
Certainly. See the referenced PR.

**Comment by jonorossi** [comment]
Merged #305. Thanks.
